message/infrastructure/sql/publisher.go

```diff
@@ -76,43 +84,61 @@ func NewPublisher(db db, config PublisherConfig) (*Publisher, error) {
 // Publish is blocking until all rows have been added to the Publisher's transaction.
 // Publisher doesn't guarantee publishing messages in a single transaction,
 // but the constructor accepts both *sql.DB and *sql.Tx, so transactions may be handled upstream by the user.
-func (p *Publisher) Publish(topic string, messages ...*message.Message) error {
+func (p *Publisher) Publish(topic string, messages ...*message.Message) (err error) {
 	if p.closed {
 		return ErrPublisherClosed
 	}
 
+	p.publishWg.Add(1)
+	defer p.publishWg.Done()
+
 	if err := validateTopicName(topic); err != nil {
 		return err
 	}
 
-	p.publishWg.Add(1)
-	defer p.publishWg.Done()
+	if err := p.initializeSchema(topic); err != nil {
+		return err
+	}
 
-	insertQuery := p.config.SchemaAdapter.InsertQuery(topic)
-	p.config.Logger.Info("Preparing query to insert messages", watermill.LogFields{
-		"q": insertQuery,
+	insertQuery, insertArgs, err := p.config.SchemaAdapter.InsertQuery(topic, messages)
+	if err != nil {
+		return errors.Wrap(err, "cannot create insert query")
+	}
+
+	p.config.Logger.Trace("Inserting message to SQL", watermill.LogFields{
+		"query":      insertQuery,
+		"query_args": sqlArgsToLog(insertArgs),
 	})
 
-	stmt, err := p.db.Prepare(insertQuery)
```

Comment:
I removed all statements, removing them made code faster (we don't need to call sql twice) and cleaner

Reply:
:+1: